Unable to install Erlang/OTP 27.1 and 26.2.5.3 on Apple Silicon macOS and Xcode 16.0

Hi,

I tried to build Erlang/OTP 27.1 on Apple Silicon macOS Sonoma and Sequoia, but got the following error:

```c
In file included from sys/unix/erl_unix_sys.h:65:
/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/unistd.h:525:10: fatal error: cannot open file 'aarch64-apple-darwin24.0.0/opt/jit/_ctermid.h': Too many open files
  525 | #include <_ctermid.h>
      |          ^
1 error generated.
make[4]: *** [obj/aarch64-apple-darwin24.0.0/opt/jit/beam_asm_module.o] Error 1
make[4]: *** Waiting for unfinished jobs....
make[3]: *** [opt] Error 2
make[2]: *** [opt] Error 2
make[1]: *** [jit] Error 2
make: *** [emulator] Error 2
```

similar issue here, but with different error:

```
In file included from aarch64-apple-darwin24.0.0/opt/jit/asmjit/./core/../core/../core/../core/globals.h:9:
aarch64-apple-darwin24.0.0/opt/jit/asmjit/./core/../core/../core/../core/../core/api-config.h:42:10: fatal error: 'iterator' file not found
   42 | #include <iterator>
   ```

Hitting similar error:

```
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/unistd.h:525:10: fatal error: cannot open file 'aarch64-apple-darwin24.0.0/opt/jit/_ctermid.h': Too many open files
  525 | #include <_ctermid.h>
      |          ^
1 error generated.
make[4]: *** [obj/aarch64-apple-darwin24.0.0/opt/jit/beam_asm_module.o] Error 1
```

Environment info:

```sh
$ which gcc
/usr/bin/gcc

$ gcc --version
Apple clang version 16.0.0 (clang-1600.0.26.3)
Target: arm64-apple-darwin24.0.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin

$ clang --version
Apple clang version 16.0.0 (clang-1600.0.26.3)
Target: arm64-apple-darwin24.0.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
```

I'm getting a different error on Apple Silicon with Xcode 16.0, any ideas?

```bash
Extracting source code
Building Erlang/OTP 27.1 (asdf_27.1), please wait...
APPLICATIONS DISABLED (See: /Users/redacted/.asdf/plugins/erlang/kerl-home/builds/asdf_27.1/otp_build_27.1.log)
 * jinterface     : No Java compiler found
 * odbc           : ODBC library - link check failed

APPLICATIONS INFORMATION (See: /Users/redacted/.asdf/plugins/erlang/kerl-home/builds/asdf_27.1/otp_build_27.1.log)
 * wx             : wxWidgets was not compiled with --enable-webview or wxWebView developer package is not installed, wxWebView will NOT be available
 * wxWidgets was not compiled with --enable-compat30, wx will NOT be useable
 * wxWidgets don't have gl support, wx will NOT be useable
 * Can not link wx program are all developer packages installed?

Build failed.
In file included from aarch64-apple-darwin24.0.0/opt/jit/asmjit/./core/../core/../core/../core/globals.h:9:
aarch64-apple-darwin24.0.0/opt/jit/asmjit/./core/../core/../core/../core/../core/api-config.h:45:10: fatal error: 'initializer_list' file not found
   45 | #include <initializer_list>
      |          ^~~~~~~~~~~~~~~~~~
1 error generated.
make[4]: *** [aarch64-apple-darwin24.0.0/opt/jit/asmjit/asmjit.hpp.gch] Error 1
make[3]: *** [opt] Error 2
make[2]: *** [opt] Error 2
make[1]: *** [jit] Error 2
make: *** [emulator] Error 2
```

> I'm getting a different error on Apple Silicon with Xcode 16.0, any ideas?
> 
> ```shell
> Extracting source code
> Building Erlang/OTP 27.1 (asdf_27.1), please wait...
> APPLICATIONS DISABLED (See: /Users/redacted/.asdf/plugins/erlang/kerl-home/builds/asdf_27.1/otp_build_27.1.log)
>  * jinterface     : No Java compiler found
>  * odbc           : ODBC library - link check failed
> 
> APPLICATIONS INFORMATION (See: /Users/redacted/.asdf/plugins/erlang/kerl-home/builds/asdf_27.1/otp_build_27.1.log)
>  * wx             : wxWidgets was not compiled with --enable-webview or wxWebView developer package is not installed, wxWebView will NOT be available
>  * wxWidgets was not compiled with --enable-compat30, wx will NOT be useable
>  * wxWidgets don't have gl support, wx will NOT be useable
>  * Can not link wx program are all developer packages installed?
> 
> Build failed.
> In file included from aarch64-apple-darwin24.0.0/opt/jit/asmjit/./core/../core/../core/../core/globals.h:9:
> aarch64-apple-darwin24.0.0/opt/jit/asmjit/./core/../core/../core/../core/../core/api-config.h:45:10: fatal error: 'initializer_list' file not found
>    45 | #include <initializer_list>
>       |          ^~~~~~~~~~~~~~~~~~
> 1 error generated.
> make[4]: *** [aarch64-apple-darwin24.0.0/opt/jit/asmjit/asmjit.hpp.gch] Error 1
> make[3]: *** [opt] Error 2
> make[2]: *** [opt] Error 2
> make[1]: *** [jit] Error 2
> make: *** [emulator] Error 2
> ```
If running `ulimit -n 65536` didn't worked. You can try this https://github.com/asdf-vm/asdf-erlang/issues/319#issuecomment-2367630445. Then don't update your xcode yet until problem is resolved.

Same for me.

On Apple M2 Pro, Sonoma 14.7, kerl 4.3.0 (without using any asdf)
```
kerl build-install 27.1.1
```
resulted in:
```
In file included from aarch64-apple-darwin23.6.0/opt/jit/asmjit/asmjit.hpp:27:
In file included from aarch64-apple-darwin23.6.0/opt/jit/asmjit/./core.h:1932:
In file included from aarch64-apple-darwin23.6.0/opt/jit/asmjit/./core/archtraits.h:9:
In file included from aarch64-apple-darwin23.6.0/opt/jit/asmjit/./core/../core/operand.h:9:
In file included from aarch64-apple-darwin23.6.0/opt/jit/asmjit/./core/../core/../core/archcommons.h:13:
In file included from aarch64-apple-darwin23.6.0/opt/jit/asmjit/./core/../core/../core/../core/globals.h:9:
aarch64-apple-darwin23.6.0/opt/jit/asmjit/./core/../core/../core/../core/../core/api-config.h:45:10: fatal error: 'initializer_list' file not found
   45 | #include <initializer_list>
      |          ^~~~~~~~~~~~~~~~~~
1 error generated.
gmake[4]: *** [aarch64-apple-darwin23.6.0/Makefile:958: aarch64-apple-darwin23.6.0/opt/jit/asmjit/asmjit.hpp.gch] Error 1
gmake[4]: Leaving directory '/Users/dansari/.kerl/builds/27.1.1/otp_src_27.1.1/erts/emulator'
gmake[3]: *** [/Users/dansari/.kerl/builds/27.1.1/otp_src_27.1.1/make/run_make.mk:35: opt] Error 2
gmake[3]: Leaving directory '/Users/dansari/.kerl/builds/27.1.1/otp_src_27.1.1/erts/emulator'
gmake[2]: *** [Makefile:45: opt] Error 2
gmake[2]: Leaving directory '/Users/dansari/.kerl/builds/27.1.1/otp_src_27.1.1/erts'
gmake[1]: *** [Makefile:60: jit] Error 2
gmake[1]: Leaving directory '/Users/dansari/.kerl/builds/27.1.1/otp_src_27.1.1/erts'
gmake: *** [Makefile:483: emulator] Error 2
```

Re-installing xcode fixed this issue.

Hello! 👋  I'm getting a different error with OTP 27.1.2 on Apple M3 Max with Sequoia:

```sh
❯ KERL_CONFIGURE_OPTIONS="--disable-parallel-configure" asdf install erlang 27.1.2

ERROR: 'asdf_27.1.2' is not a kerl-managed Erlang/OTP installation.
Build 'asdf_27.1.2' has been deleted.
Extracting source code for normal build...
Building (normal) Erlang/OTP 27.1.2 (asdf_27.1.2); please wait...
Initializing (build) log file at /Users/imcosta/.asdf/plugins/erlang/kerl-home/builds/asdf_27.1.2/otp_build_27.1.2.log.
Attempting to use Homebrew OpenSSL from /usr/local/opt/openssl@3.0...
... found!
APPLICATIONS DISABLED (See: /Users/imcosta/.asdf/plugins/erlang/kerl-home/builds/asdf_27.1.2/otp_build_27.1.2.log)
 * jinterface     : No Java compiler found
 * odbc           : ODBC library - link check failed

APPLICATIONS INFORMATION (See: /Users/imcosta/.asdf/plugins/erlang/kerl-home/builds/asdf_27.1.2/otp_build_27.1.2.log)
 * wx             : wxWidgets was not compiled with --enable-webview or wxWebView developer package is not installed, wxWebView will NOT be available
 * Can not link wx program are all developer packages installed?

ERROR: build failed.
=== Leaving application mnesia
=== Entering application crypto
make[3]: Nothing to be done for `opt'.
 MAKE   opt
make[4]: aarch64-apple-darwin24.0.0/Makefile: No such file or directory
make[4]: *** No rule to make target `aarch64-apple-darwin24.0.0/Makefile'.  Stop.
make[3]: *** [opt] Error 2
make[2]: *** [opt] Error 2
make[1]: *** [opt] Error 2
make: *** [libs] Error 2
```

**I tried different things:**
- Reinstall Xcode Command Line Tools following [this](https://github.com/asdf-vm/asdf-erlang/issues/319#issuecomment-2367630445)
- Change [ulimits](https://github.com/asdf-vm/asdf-erlang/issues/319#issuecomment-2367727142)
- Use `CPP=cpp EGREP=egrep KERL_CONFIGURE_OPTIONS="--enable-darwin-64bit --disable-parallel-configure"`
- Remove the `--disable-parallel-configure` but in that case I got a different error:

```sh
checking for OpenSSL header in /usr/local/opt/openssl@3.0... yes
checking for OpenSSL in /usr/local/opt/openssl@3.0... configure: error: neither static nor dynamic crypto library found in /usr/local/opt/openssl@3.0
ERROR: /Users/imcosta/.asdf/plugins/erlang/kerl-home/builds/asdf_27.1.2/otp_src_27.1.2/lib/crypto/configure failed!
```
At this point I don't know what to do. Any ideas? Thanks!

--- a/bin/install
+++ b/bin/install
@@ -110,7 +110,8 @@ set_unixodbc_opt() {
         fi
 
         if [[ ! "$KERL_CONFIGURE_OPTIONS" =~ unixodbc ]]; then
-            local kerl_unixodbc_opt; kerl_unixodbc_opt=" --with-odbc=$(brew --prefix unixodbc)";
+            local kerl_unixodbc_opt
+            kerl_unixodbc_opt=" --with-odbc=$(brew --prefix unixodbc)"
             export KERL_CONFIGURE_OPTIONS+=" $kerl_unixodbc_opt"
             echo "[asdf-erlang] 🛟 Added unixodbc to KERL_CONFIGURE_OPTIONS: $kerl_unixodbc_opt"
         fi
@@ -136,7 +137,8 @@ set_unixodbc_opt() {
             echo "[asdf-erlang] 🛟 No LDFLAGS found. Setting LDFLAGS to: $LDFLAGS"
             LDFLAGS_SET_BY_ASDF_ERLANG=1 # so that we can clear it later.
         elif [[ "$LDFLAGS" != *unixodbc* ]]; then
-            local unixodbc_lib_path; unixodbc_lib_path="$(brew --prefix unixodbc)/lib"
+            local unixodbc_lib_path
+            unixodbc_lib_path="$(brew --prefix unixodbc)/lib"
             export LDFLAGS+=" -L$unixodbc_lib_path"
             echo "[asdf-erlang] 🛟 Added $unixodbc_lib_path to LDFLAGS env var"
         fi
@@ -149,7 +151,8 @@ add_openjdk_to_path() {
             return
         fi
 
-        local openjdk_path; openjdk_path="$(brew --prefix openjdk)"
+        local openjdk_path
+        openjdk_path="$(brew --prefix openjdk)"
         if [[ ":$PATH:" != *":$openjdk_path/bin:"* ]]; then
             export PATH="$openjdk_path/bin:$PATH"
             echo "[asdf-erlang] 🛟 OpenJDK has been added to PATH for this terminal session: $openjdk_path/bin"
@@ -193,7 +196,8 @@ set_ssl_opt() {
 
         # If openssl is not already in KERL_CONFIGURE_OPTIONS, then add it.
         if [[ ! "$KERL_CONFIGURE_OPTIONS" =~ openssl ]]; then
-            local kerl_ssl_opt; kerl_ssl_opt=" --with-ssl=$(brew --prefix openssl@3)";
+            local kerl_ssl_opt
+            kerl_ssl_opt=" --with-ssl=$(brew --prefix openssl@3)"
             export KERL_CONFIGURE_OPTIONS+=" $kerl_ssl_opt"
             echo "[asdf-erlang] 🛟 Added openssl to KERL_CONFIGURE_OPTIONS: $kerl_ssl_opt"
         fi

--- a/bin/install
+++ b/bin/install
@@ -110,7 +110,7 @@ set_unixodbc_opt() {
         fi
 
         if [[ ! "$KERL_CONFIGURE_OPTIONS" =~ unixodbc ]]; then
-            local kerl_unixodbc_opt=" --with-odbc=$(brew --prefix unixodbc)"
+            local kerl_unixodbc_opt; kerl_unixodbc_opt=" --with-odbc=$(brew --prefix unixodbc)";
             export KERL_CONFIGURE_OPTIONS+=" $kerl_unixodbc_opt"
             echo "[asdf-erlang] 🛟 Added unixodbc to KERL_CONFIGURE_OPTIONS: $kerl_unixodbc_opt"
         fi
@@ -118,22 +118,25 @@ set_unixodbc_opt() {
         # If CC not set, then set CC.
         # ELSE add the unixodbc include path to CC.
         if [ -z "${CC:-}" ]; then
-            export CC="/usr/bin/clang -I$(brew --prefix unixodbc)/include"
+            CC="/usr/bin/clang -I$(brew --prefix unixodbc)/include"
+            export CC
             echo "[asdf-erlang] 🛟 No CC found. Setting CC to: $CC"
             CC_SET_BY_ASDF_ERLANG=1 # so that we can clear it later.
         elif [[ "$CC" != *unixodbc* ]]; then
-            export CC+=" -I$(brew --prefix unixodbc)/include"
+            CC+=" -I$(brew --prefix unixodbc)/include"
+            export CC
             echo "[asdf-erlang] 🛟 Added unixodbc include path to CC: $CC"
         fi
 
         # If LDFLAGS not set, then set LDFLAGS.
         # ELSE add the unixodbc library path to LDFLAGS.
         if [ -z "${LDFLAGS:-}" ]; then
-            export LDFLAGS="-L$(brew --prefix unixodbc)/lib"
+            LDFLAGS="-L$(brew --prefix unixodbc)/lib"
+            export LDFLAGS
             echo "[asdf-erlang] 🛟 No LDFLAGS found. Setting LDFLAGS to: $LDFLAGS"
             LDFLAGS_SET_BY_ASDF_ERLANG=1 # so that we can clear it later.
         elif [[ "$LDFLAGS" != *unixodbc* ]]; then
-            local unixodbc_lib_path="$(brew --prefix unixodbc)/lib"
+            local unixodbc_lib_path; unixodbc_lib_path="$(brew --prefix unixodbc)/lib"
             export LDFLAGS+=" -L$unixodbc_lib_path"
             echo "[asdf-erlang] 🛟 Added $unixodbc_lib_path to LDFLAGS env var"
         fi
@@ -146,7 +149,7 @@ add_openjdk_to_path() {
             return
         fi
 
-        local openjdk_path="$(brew --prefix openjdk)"
+        local openjdk_path; openjdk_path="$(brew --prefix openjdk)"
         if [[ ":$PATH:" != *":$openjdk_path/bin:"* ]]; then
             export PATH="$openjdk_path/bin:$PATH"
             echo "[asdf-erlang] 🛟 OpenJDK has been added to PATH for this terminal session: $openjdk_path/bin"
@@ -190,7 +193,7 @@ set_ssl_opt() {
 
         # If openssl is not already in KERL_CONFIGURE_OPTIONS, then add it.
         if [[ ! "$KERL_CONFIGURE_OPTIONS" =~ openssl ]]; then
-            local kerl_ssl_opt=" --with-ssl=$(brew --prefix openssl@3)"
+            local kerl_ssl_opt; kerl_ssl_opt=" --with-ssl=$(brew --prefix openssl@3)";
             export KERL_CONFIGURE_OPTIONS+=" $kerl_ssl_opt"
             echo "[asdf-erlang] 🛟 Added openssl to KERL_CONFIGURE_OPTIONS: $kerl_ssl_opt"
         fi

--- a/bin/install
+++ b/bin/install
@@ -13,12 +13,31 @@ install_erlang() {
     ensure_kerl_setup
     local build_name
 
+    install_dependency_checks
+    ensure_kerl_config_opts
+    set_unixodbc_opt
+    set_ssl_opt
+    add_openjdk_to_path
+    ensure_ulimit
+
+    echo "[asdf-erlang] 📦 Building with $(env | grep KERL_CONFIGURE_OPTIONS)"
+
     build_name="asdf_$ASDF_INSTALL_VERSION"
 
     export MAKEFLAGS="-j$ASDF_CONCURRENCY"
 
-    $(kerl_path) delete installation "$build_name" || true
-    $(kerl_path) delete build "$build_name" || true
+    if $(kerl_path) list installations | grep -q "$build_name"; then
+        echo "[asdf-erlang] 🧹 Cleanup kerl installation $build_name"
+        $(kerl_path) delete installation "$build_name" || true
+    else
+        echo "[asdf-erlang] ❄️ No kerl installation to cleanup for $build_name"
+    fi
+    if $(kerl_path) list builds | grep -q "$build_name"; then
+        echo "[asdf-erlang] 🧹 Cleanup kerl build $build_name"
+        $(kerl_path) delete build "$build_name" || true
+    else
+        echo "[asdf-erlang] ❄️ No kerl build to cleanup for $build_name"
+    fi
 
     if [ "$ASDF_INSTALL_TYPE" = "ref" ]; then
         $(kerl_path) build git "${OTP_GITHUB_URL:-https://github.com/erlang/otp.git}" "$ASDF_INSTALL_VERSION" "$build_name"
@@ -29,10 +48,16 @@ install_erlang() {
     # We hide all output from this command so the
     # "You can activate this installation running the following command:"
     # that doesn't apply is hidden
+    echo "🧹 Cleanup kerl install $build_name $ASDF_INSTALL_PATH"
     $(kerl_path) install "$build_name" "$ASDF_INSTALL_PATH" >/dev/null 2>&1
+
+    echo "🧹 Cleanup kerl build $build_name"
     $(kerl_path) cleanup "$build_name"
 
+    echo "🔗 Linking app executables $ASDF_INSTALL_PATH"
     link_app_executables "$ASDF_INSTALL_PATH"
+    cleanup_custom_env_vars
+    echo "👍 Installed erlang ${ASDF_INSTALL_VERSION}"
 }
 
 link_app_executables() {
@@ -51,6 +76,135 @@ link_app_executables() {
             ln -s "$file" .
         fi
     done
+}
+
+install_dependency_checks() {
+    if [[ "$OSTYPE" == "darwin"* ]]; then
+        installed_packages=()
+        for package in fop openssl unixodbc openjdk wxmac; do
+            if brew list --versions "$package" >/dev/null 2>&1; then
+                installed_packages+=("$package")
+            else
+                if [ "$package" = "wxmac" ]; then
+                    echo "[asdf-erlang] ⚠️ Warning: $package is optional and not installed. You can install it using 'brew install --build-from-source wxmac'."
+                    echo "[asdf-erlang] ⚠️ Note: wxmac is required for building Erlang/OTP with a working :observer"
+                else
+                    echo "[asdf-erlang] ⚠️ Warning: $package is optional and not installed. Please install it using 'brew install $package'."
+                fi
+            fi
+        done
+    fi
+}
+
+ensure_kerl_config_opts() {
+    if [[ -z "${KERL_CONFIGURE_OPTIONS:-}" ]]; then
+        export KERL_CONFIGURE_OPTIONS=""
+    fi
+}
+
+set_unixodbc_opt() {
+    if [[ "$OSTYPE" == "darwin"* ]]; then
+        # If no unixodbc is installed, then skip.
+        if ! brew --prefix unixodbc >/dev/null 2>&1; then
+            return
+        fi
+
+        if [[ ! "$KERL_CONFIGURE_OPTIONS" =~ unixodbc ]]; then
+            local kerl_unixodbc_opt=" --with-odbc=$(brew --prefix unixodbc)"
+            export KERL_CONFIGURE_OPTIONS+=" $kerl_unixodbc_opt"
+            echo "[asdf-erlang] 🛟 Added unixodbc to KERL_CONFIGURE_OPTIONS: $kerl_unixodbc_opt"
+        fi
+
+        # If CC not set, then set CC.
+        # ELSE add the unixodbc include path to CC.
+        if [ -z "${CC:-}" ]; then
+            export CC="/usr/bin/clang -I$(brew --prefix unixodbc)/include"
+            echo "[asdf-erlang] 🛟 No CC found. Setting CC to: $CC"
+            CC_SET_BY_ASDF_ERLANG=1 # so that we can clear it later.
+        elif [[ "$CC" != *unixodbc* ]]; then
+            export CC+=" -I$(brew --prefix unixodbc)/include"
+            echo "[asdf-erlang] 🛟 Added unixodbc include path to CC: $CC"
+        fi
+
+        # If LDFLAGS not set, then set LDFLAGS.
+        # ELSE add the unixodbc library path to LDFLAGS.
+        if [ -z "${LDFLAGS:-}" ]; then
+            export LDFLAGS="-L$(brew --prefix unixodbc)/lib"
+            echo "[asdf-erlang] 🛟 No LDFLAGS found. Setting LDFLAGS to: $LDFLAGS"
+            LDFLAGS_SET_BY_ASDF_ERLANG=1 # so that we can clear it later.
+        elif [[ "$LDFLAGS" != *unixodbc* ]]; then
+            local unixodbc_lib_path="$(brew --prefix unixodbc)/lib"
+            export LDFLAGS+=" -L$unixodbc_lib_path"
+            echo "[asdf-erlang] 🛟 Added $unixodbc_lib_path to LDFLAGS env var"
+        fi
+    fi
+}
+
+add_openjdk_to_path() {
+    if [[ "$OSTYPE" == "darwin"* ]]; then
+        if ! brew --prefix openjdk >/dev/null 2>&1; then
+            return
+        fi
+
+        local openjdk_path="$(brew --prefix openjdk)"
+        if [[ ":$PATH:" != *":$openjdk_path/bin:"* ]]; then
+            export PATH="$openjdk_path/bin:$PATH"
+            echo "[asdf-erlang] 🛟 OpenJDK has been added to PATH for this terminal session: $openjdk_path/bin"
+            echo "[asdf-erlang] Please ensure this is included in your shell's dot files (.zshrc, .bashrc, etc.)"
+        fi
+    fi
+}
+
+ensure_ulimit() {
+    if [ "$(ulimit -n)" -lt 1000 ]; then
+        ulimit -n 65536
+        echo "[asdf-erlang] 🛟 ulimit was low. It has been set to 65536 for this terminal session"
+    fi
+}
+
+set_ssl_opt() {
+    if [[ "$OSTYPE" == "darwin"* ]]; then
+        # If this is a ref install, then we don't need to set the ssl option.
+        # Reason: Don't want to handle that for now.
+        if [[ "$ASDF_INSTALL_TYPE" == "ref" ]]; then
+            echo "[asdf-erlang] ⚠️ Skipping setting --with-ssl in KERL_CONFIGURE_OPTIONS for ref install"
+            return
+        fi
+
+        otp_major=$(echo "$ASDF_INSTALL_VERSION" | cut -d. -f1)
+        otp_minor=$(echo "$ASDF_INSTALL_VERSION" | cut -d. -f2)
+
+        # Copied IF condition check from the kerl binary to match their version check.
+        # We only have to fix/handle newer erlang+openssl versions because of kerl looking for openssl@3.0 instead of openssl#3.
+        # The erlang version that use openssl 1.1 should be fine (or as is).
+        if [ "$otp_major" = 'git' ] || [ "$otp_major" -lt 25 ] || { [ "$otp_major" -eq 25 ] && [ "$otp_minor" -lt 1 ]; }; then
+            echo "[asdf-erlang] ⚠️ Skipping setting --with-ssl in KERL_CONFIGURE_OPTIONS. This erlang version uses openssl v1.x"
+            return
+        fi
+
+        # Only set the ssl option for newer erlang versions that use openssl@3.
+        if ! brew --prefix openssl@3 >/dev/null 2>&1; then
+            echo "[asdf-erlang] ⚠️ Skipping setting --with-ssl in KERL_CONFIGURE_OPTIONS. brew prefix path for openssl@3 not found."
+            return
+        fi
+
+        # If openssl is not already in KERL_CONFIGURE_OPTIONS, then add it.
+        if [[ ! "$KERL_CONFIGURE_OPTIONS" =~ openssl ]]; then
+            local kerl_ssl_opt=" --with-ssl=$(brew --prefix openssl@3)"
+            export KERL_CONFIGURE_OPTIONS+=" $kerl_ssl_opt"
+            echo "[asdf-erlang] 🛟 Added openssl to KERL_CONFIGURE_OPTIONS: $kerl_ssl_opt"
+        fi
+    fi
+}
+
+cleanup_custom_env_vars() {
+    if [[ "${LDFLAGS_SET_BY_ASDF_ERLANG:-}" == "1" ]]; then
+        unset LDFLAGS
+    fi
+
+    if [[ "${CC_SET_BY_ASDF_ERLANG:-}" == "1" ]]; then
+        unset CC
+    fi
 }
 
 install_erlang

